x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/conditional/SpatialRelatesFunction.java

```diff
@@ -0,0 +1,580 @@
+            }
+        }
+    }
+
+    protected class SpatialEvaluatorWithConstantFactory extends SpatialEvaluatorFactory<
+        EvalOperator.ExpressionEvaluator.Factory,
+        Component2D> {
+
+        SpatialEvaluatorWithConstantFactory(
+            TriFunction<
+                Source,
+                EvalOperator.ExpressionEvaluator.Factory,
+                Component2D,
+                EvalOperator.ExpressionEvaluator.Factory> factoryCreator
+        ) {
+            super(factoryCreator);
+        }
+
+        @Override
+        public EvalOperator.ExpressionEvaluator.Factory get(Function<Expression, EvalOperator.ExpressionEvaluator.Factory> toEvaluator) {
+            if (swapSides) {
+                return factoryCreator.apply(source(), toEvaluator.apply(right()), asLuceneComponent2D(crsType, left()));
+            } else {
+                return factoryCreator.apply(source(), toEvaluator.apply(left()), asLuceneComponent2D(crsType, right()));
+            }
+        }
+    }
+
+    protected class SpatialEvaluatorWithConstantsFactory extends SpatialEvaluatorFactory<GeometryDocValueReader, Component2D> {
+
+        SpatialEvaluatorWithConstantsFactory(
+            TriFunction<Source, GeometryDocValueReader, Component2D, EvalOperator.ExpressionEvaluator.Factory> factoryCreator
+        ) {
+            super(factoryCreator);
+        }
+
+        @Override
+        public EvalOperator.ExpressionEvaluator.Factory get(Function<Expression, EvalOperator.ExpressionEvaluator.Factory> toEvaluator) {
+            try {
+                if (swapSides) {
+                    return factoryCreator.apply(source(), asGeometryDocValueReader(crsType, right()), asLuceneComponent2D(crsType, left()));
+                } else {
+                    return factoryCreator.apply(source(), asGeometryDocValueReader(crsType, left()), asLuceneComponent2D(crsType, right()));
+                }
+            } catch (IOException e) {
+                throw new IllegalArgumentException("Failed to process spatial constant", e);
+            }
+        }
+    }
+
+    protected enum SpatialCrsTypes {
+        GEO,
+        CARTESIAN,
+        UNSPECIFIED;
+
+        public static SpatialCrsTypes fromDataType(DataType dataType) {
+            return EsqlDataTypes.isSpatialGeo(dataType) ? SpatialCrsTypes.GEO
+                : EsqlDataTypes.isSpatial(dataType) ? SpatialCrsTypes.CARTESIAN
+                : SpatialCrsTypes.UNSPECIFIED;
+        }
+    }
+
+    protected record SpatialEvaluatorFieldKey(DataType dataType, boolean isConstant) {}
+
+    protected record SpatialEvaluatorKey(
+        SpatialCrsTypes crsType,
+        boolean useDocValues,
+        SpatialEvaluatorFieldKey left,
+        SpatialEvaluatorFieldKey right
+    ) {
+        SpatialEvaluatorKey withDocValues() {
+            return new SpatialEvaluatorKey(crsType, true, left, right);
+        }
+
+        SpatialEvaluatorKey swapSides() {
```

Comment:
The QL infrastructure already has a fair bit of stuff around constants and rotating constants to the right hand side of an expression. Is that something we can link into here?

Reply:
I also think we can do something like what is done in the `AggregateMapper`, where all the various combinations are considered upfront for all aggregations (and a few special cases, like Spatial that I added). I think when I code the next spatial aggregation, and find the type rules are identical to this one, I'll be very tempted to do just that.

Reply:
I've improved this in the work to make the mappers static, but it does not yet move this out to whatever QL infrastructure supports this.